cmd/csi-node-driver-registrar/node_register.go

```diff
@@ -63,6 +65,19 @@ func nodeRegister(csiDriverName, httpEndpoint string) {
 	// Registers kubelet plugin watcher api.
 	registerapi.RegisterRegistrationServer(grpcServer, registrar)
 
+	// Sometimes on windows after registration with the kubelet plugin we don't
+	// get a callback through GetInfo, as a workaround if we don't get a callback within
+	// the next 10 seconds we'll restart
+	go func() {
+		err := wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
```

Comment:
I could adjust the timeout based on how soon we get a callback, the timing in my dev cluster was:

- linux: around 2s
- windows: around 1.7s

Reply:
In general, I think its's preferable to use a readiness probe so that the behavior and thresholds can be configurable by users. 

However, http probes don't work well if the pod needs to run on host network (which most csi node plugins have to do). An exec probe is better in that case.

Reply:
good to know this, yes another idea that I posted on #143 was to use probes however I didn't know what to check on the probe, on successful registration we could `touch` a file that could be checked for existence with the exec probe but at the same time cleaning that file could generate more problems too, I could connect to the api-server to ask if this component is registered too but this would probably need RBAC rules and so on.

Another idea was to use a flag in the CLI for the timeout, because the state checked is only local I believe that checking every 100ms would be fine, there are two flags to control timeouts (one deprecated) and we could have another one for the kubelet registration timeout:

```
	connectionTimeout       = flag.Duration("connection-timeout", 0, "The --connection-timeout flag is deprecated")
	operationTimeout        = flag.Duration("timeout", time.Second, "Timeout for waiting for communication with driver")

	// new flag
	kubeletRegistrationTimeout        = flag.Duration("kubelet-registration-timeout", 5*time.Second, "Timeout for waiting kubelet registration")
```

Reply:
Can the file be put into an emptydir directory? That will get cleaned up when the pod gets deleted.

Reply:
just want to understand, comparing this current implementation and "touch file" approach, what are pros and cons?
this one seems simpler?

Reply:
I think that the current implementation is simpler and self contained (no additional setup on the yaml), however as Michelle said it lacks a way to control the timeout which we could do with a cli flag too.

For the touch file approach in a experiment I deployed this component (that timed out after 10s)  together the GCE PD CSI driver and if I remember correctly only this container was restarted, I'm going to deploy the version that crashed the container again and see if the pods get restarted.

Reply:
it's 3 things: timeout, polling frequency, and whether or not you even want to turn on the behavior.

Reply:
I checked that only the node-driver-registrar container is restarted:

![image](https://user-images.githubusercontent.com/1616682/123150826-a7d4a780-d417-11eb-85b6-f24007e345f3.png)
![image](https://user-images.githubusercontent.com/1616682/123151075-eff3ca00-d417-11eb-8551-793cbd0204f0.png)

Also the docs say:

> Note: A container crashing does not remove a Pod from a node. The data in an emptyDir volume is safe across container crashes.

The touch file approach would be something like this:

```
func (e registrationServer) GetInfo(ctx context.Context, req *registerapi.InfoRequest) (*registerapi.PluginInfo, error) {
	klog.Infof("Received GetInfo call: %+v", req)
	// TODO: touch file /kubelet-registration-ack/ack
	return &registerapi.PluginInfo{
		Type:              registerapi.CSIPlugin,
		Name:              e.driverName,
    }
}

func main() {
   // TODO: add trap to delete the temp file
}
```

```
        - name: csi-driver-registrar
          image: gcr.io/mauriciopoppe-gke-dev/csi-node-driver-registrar:canary
          imagePullPolicy: Always
          args:
            - --v=5
            - --csi-address=unix://C:\\csi\\csi.sock
            - --kubelet-registration-path=C:\\var\\lib\\kubelet\\plugins\\pd.csi.storage.gke.io\\csi.sock
          env:
            - name: KUBE_NODE_NAME
              valueFrom:
                fieldRef:
                  fieldPath: spec.nodeName
          volumeMounts:
            - name: kubelet-registration-ack-dir
              mountPath: /kubelet-registration-ack
          livenessProbe:
            exec:
              command:
              - cat
              - /kubelet-registration-ack/ack
            initialDelaySeconds: 1
            periodSeconds: 1
```

A problem that I see is if node-driver-registrar gets forcefully restarted without cleaning the temp file, the next time it comes up it could think that it's alive because the file is there but it could be possible that it didn't perform the registration process correctly.

Reply:
Can you delete the file on startup?

One more challenge is that the node-driver-registrar container doesn't have a shell. So it may need to be a command line option to do the readiness check.

Reply:
That approach sounds good, I thought about creating it with the name of the driver e.g.

```
          args:
            - --v=5
            - --csi-address=unix://C:\\csi\\csi.sock
            - --kubelet-registration-ack-path=C:\\registration\\pd.csi.storage.gke.io-kubelet-registration-ack
            - --kubelet-registration-path=C:\\var\\lib\\kubelet\\plugins\\pd.csi.storage.gke.io\\csi.sock
          livenessProbe:
            exec:
              command:
              - /node-driver-registrar --kubelet-registration-check-ack=C:\\registration\\pd.csi.storage.gke.io-kubelet-registration-ack
            initialDelaySeconds: 1
            periodSeconds: 1
```

There would be 2 additional flags:

- `--kubelet-registration-ack-path` that tells this component where to write the lock file once it gets a callback from the kubelet
- `--kubelet-registration-check-ack` with the same value as kubelet-registration-ack-path

The name of driver could be fetched from the driver itself however it makes it difficult during a forced restart when we have to delete the file on startup because we don't know which file to delete until the gRPC request to driver is made, that's why I opted for a 2 flag approach.